src/client/utils/ajax/ajax.ts

```diff
@@ -69,17 +70,26 @@ export class Ajax {
       const timezone = env ? env.timezone : null;
       const data = { dataCube: dataCubeName, expression: ex.toJS(), timezone };
       return Ajax.query<{ result: DatasetJS }>({ method, url, timeout, data }, oauth)
-        .then(res => Dataset.fromJS(res.result));
+          .then(res => Dataset.fromJS(res.result));
     };
   }
 
-  static sources(appSettings: ClientAppSettings): Promise<ClientSources> {
-    const headers = Ajax.headers(appSettings.oauth);
-    return axios.get<SerializedSources>("sources", { headers })
-      .then(resp => resp.data)
-      .catch(error => {
-        throw mapOauthError(appSettings.oauth, error);
-      })
-      .then(sourcesJS => deserialize(sourcesJS, appSettings));
+  static async sources(appSettings: ClientAppSettings): Promise<ClientSources> {
+    try {
+      const headers = Ajax.headers(appSettings.oauth);
+      const clusters = (await axios.get("sources/clusters", { headers })).data;
+      let dataCubesResult: SerializedDataCube[] = [];
+      let isDone = false;
+      let batch = 0;
+      while (!isDone) {
```

Comment:
I'm wondering if we could extract this to async iterator and with `for await ... of`. That's very loose idea though!